### PR TITLE
feat: Authenticate maker RPC requests with a local cookie

### LINKED ISSUE #877: [Loupe Finding] feat: Authenticate maker RPC requests with a local cookie token
## Problem Statement
`makerd` exposes a localhost TCP RPC control plane used by maker-cli for wallet and server operations. The current protocol accepts raw CBOR `RpcMsgReq` messages without authentication, so any local process or reachable peer that can connect to the RPC port can issue privileged commands such as wallet balance/UTXO queries, address generation, daemon shutdown, wallet sync, or `SendToAddress`.

## Proposed Solution
Add daemon-owned RPC authentication by generating a high-entropy local cookie in the maker data directory with restrictive file permissions, requiring clients to wrap each RPC request with that token, and rejecting unauthenticated or invalid-token requests before dispatching privileged operations. 
`maker-cli` should load the cookie from the matching maker data directory and include it transparently in every request.

## Affected Component(s)
- [x] **makerd** (background daemon)
- [ ] **taker** (CLI client)
- [ ] **maker-cli** (command-line tool)
- [ ] Core protocol / cryptography / library
- [ ] **watch_tower** (contract monitoring & recovery)
- [ ] Docker / deployment scripts
- [ ] Tests / test framework
- [ ] Documentation (`docs/`)
- [ ] Other (please specify):

## Protocol Version
- [ ] Legacy (ECDSA)
- [ ] Taproot-Musig2
- [x] Both
- [ ] New protocol addition
- [ ] Not applicable

### PR DESCRIPTION
**Pull request by 0xEgao**
# Pull Request

## Description
 Adds cookie-based authentication to the `makerd` RPC control plane. `makerd` generates a fresh 256-bit cookie with owner-only permissions, while `maker-cli` reads it from the matching maker data directory and includes it with every request. Requests with missing or invalid authentication are rejected before privileged command dispatch. This keeps the change focused on authentication without introducing a new RPC API or configuration layer.

## Related Issue(s)
Fixes #877 

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactor / performance improvement
- [ ] Documentation update only
- [ ] CI / Docker / Build changes
- [ ] Other (please describe):

## Protocol Version(s) Affected
- [ ] Legacy (ECDSA — `messages.rs`, `contract.rs`, `handlers.rs`)
- [ ] Taproot-Musig2 (`messages2.rs`, `contract2.rs`, `handlers2.rs`)
- [ ] Both
- [x] Neither (infrastructure / tooling only)

> **Note:** When modifying protocol logic, changes must be made to **both** versions unless the
> scope is explicitly version-specific.

## Affected Component(s)
- [ ] **makerd** (background daemon)
- [ ] **taker** (CLI client)
- [x] **maker-cli** (command-line tool)
- [ ] Core protocol / cryptography / library
- [ ] **watch_tower** (contract monitoring & recovery)
- [ ] Docker / deployment scripts
- [ ] Tests / test framework
- [ ] Documentation (`docs/`)
- [ ] Other (please specify):

### Code Quality
- [x] I ran `cargo +nightly fmt --all` and committed the result
- [x] I ran `cargo +stable clippy --all-features --lib --bins --tests -- -D warnings` with zero warnings
- [x] I ran `cargo +stable clippy --examples -- -D warnings` with zero warnings
- [ ] I ran `RUSTDOCFLAGS="-D warnings" cargo +nightly doc --all-features --document-private-items --no-deps` with zero warnings
- [ ] Pre-commit git hook passes (`ln -s ../../git_hooks/pre-commit .git/hooks/pre-commit` if not already set)

### Testing
- [x] All unit tests pass (`cargo test`)
- [x] Integration tests pass (`cargo test --features integration-test`)
- [x] Changes were manually tested on **signet**
- [x] End-to-end maker ↔ taker swap tested (where applicable)
- [x] Test-only code is gated behind `#[cfg(feature = "integration-test")]`

### Documentation
- [ ] Relevant files in the `docs/` folder were updated
- [ ] Complex logic is commented

### Security & Privacy (Critical)
- [x] This change preserves **trustlessness** and **atomic swap guarantees**
- [x] No regression in **sybil resistance** or **fidelity bond** logic
- [x] Tor anonymity and P2P message flow were reviewed
- [x] No new attack vectors or trust assumptions introduced
- [x] Edge cases and error handling considered
- [x] ZMQ subscription integrity (raw tx/block feed) is unaffected
- [x] `integration-test` feature flag is not reachable in production code paths

## Summary by CodeRabbit

* **New Features**
  * Added cookie-based authentication for local RPC calls made by the CLI.
  * Added a `--data-directory` option to control where the RPC authentication cookie is read from.
  * RPC servers now generate, persist, and use a unique authentication cookie for connection validation.
* **Bug Fixes**
  * Unauthenticated RPC requests are now rejected with an authorization error.
  * Cookie files are written with restricted permissions to protect the token.
* **Tests**
  * Added coverage for cookie uniqueness, matching, and (on Unix) secure file permissions.

### CODE REVIEW

File: src/maker/rpc/server.rs [hunk 1]
```diff
@@ -21,6 +25,41 @@ use crate::{
...
+        use std::os::unix::fs::OpenOptionsExt;
+        options.mode(0o600);
+    }
+
+    let mut file = options.open(data_dir.join(RPC_COOKIE_FILE))?;
+    #[cfg(unix)]
+    {
+        use std::os::unix::fs::PermissionsExt;
+        file.set_permissions(fs::Permissions::from_mode(0o600))?;
+    }
```
[the remark this answers is not in the document]
Comment (0xEgao) [reply src/maker/rpc/server.rs:48]: Not required right now

**Review by 6D-pixel — COMMENTED** [review]
.

**Review by Godzilla-oss — COMMENTED** [review]
Looks good. This adds a restriction at the user level for accessing the cookies. Is it possible to set permission only to Coinswap programme to read cookies in Unix? or something similar.

**Review by 0xEgao — COMMENTED** [review]
> Looks good. This adds a restriction at the user level for accessing the cookies. Is it possible to set permission only to Coinswap programme to read cookies in Unix? or something similar.

I don’t think that’s possible with standard Unix file permissions, which operate at the user and group level rather than the program level. Program-specific access would require system-level enforcement such as AppArmor or SELinux, which is outside this PR’s scope. I think `0600` is the best option here.

File: src/maker/rpc/server.rs [hunk 2]
```diff
@@ -30,9 +69,24 @@ pub trait MakerRpc {
...
+    rpc_cookie: &str,
+) -> Result<(), MakerError> {
     let msg_bytes = read_message(socket)?;
-    let rpc_request: RpcMsgReq = serde_cbor::from_slice(&msg_bytes)?;
+    let envelope = match serde_cbor::from_slice::<AuthenticatedRpcRequest>(&msg_bytes) {
+        Ok(envelope) if cookie_matches(&envelope.token, rpc_cookie) => envelope,
+        _ => {
+            log::warn!(
+                "Rejected unauthenticated RPC request from {:?}",
+                socket.peer_addr()
```
Comment (stark-3k) [inline src/maker/rpc/server.rs]: we can use a unwrap here

**Review by stark-3k — CHANGES_REQUESTED** [review]
we can also clean up the cookie on shutdown

**Comment by 0xEgao** [comment]
> we can also clean up the cookie on shutdown

Done, now the cookie gets deleted on shutdown, and a fresh one is created on every RPC server startup

**Review by stark-3k — APPROVED** [review]
Ack
